Rubberduck.CodeAnalysis/Inspections/Abstract/DeclarationInspectionUsingGlobalInformationBaseBase.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Rubberduck.Parsing.Inspections.Abstract;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.Parsing.VBA.DeclarationCaching;
+using Rubberduck.VBEditor;
+
+namespace Rubberduck.Inspections.Abstract
+{
+    public abstract class DeclarationInspectionUsingGlobalInformationBaseBase<T> : InspectionBase
+    {
+        protected readonly DeclarationType[] RelevantDeclarationTypes;
+        protected readonly DeclarationType[] ExcludeDeclarationTypes;
+
+        protected DeclarationInspectionUsingGlobalInformationBaseBase(RubberduckParserState state, params DeclarationType[] relevantDeclarationTypes)
+            : base(state)
+        {
+            RelevantDeclarationTypes = relevantDeclarationTypes;
+            ExcludeDeclarationTypes = new DeclarationType[0];
+        }
+
+        protected DeclarationInspectionUsingGlobalInformationBaseBase(RubberduckParserState state, DeclarationType[] relevantDeclarationTypes, DeclarationType[] excludeDeclarationTypes)
+            : base(state)
+        {
+            RelevantDeclarationTypes = relevantDeclarationTypes;
+            ExcludeDeclarationTypes = excludeDeclarationTypes;
+        }
+
+        protected abstract IEnumerable<IInspectionResult> DoGetInspectionResults(QualifiedModuleName module, DeclarationFinder finder, T globalInformation);
+        protected abstract T GlobalInformation(DeclarationFinder finder);
+
+        protected virtual T GlobalInformation(QualifiedModuleName module, DeclarationFinder finder)
+        {
+            return GlobalInformation(finder);
+        }
+
+        protected override IEnumerable<IInspectionResult> DoGetInspectionResults()
+        {
+            var finder = DeclarationFinderProvider.DeclarationFinder;
+            var globalInformation = GlobalInformation(finder);
+
+            var results = new List<IInspectionResult>();
+            foreach (var moduleDeclaration in finder.UserDeclarations(DeclarationType.Module))
+            {
+                if (moduleDeclaration == null)
+                {
+                    continue;
+                }
+
+                var module = moduleDeclaration.QualifiedModuleName;
+                results.AddRange(DoGetInspectionResults(module, finder, globalInformation));
+            }
+
+            foreach (var projectDeclaration in finder.UserDeclarations(DeclarationType.Project))
+            {
+                if (projectDeclaration == null)
+                {
+                    continue;
+                }
+
+                var module = projectDeclaration.QualifiedModuleName;
+                results.AddRange(DoGetInspectionResults(module, finder, globalInformation));
```

Comment:
This does not account for the option of module-dependent global information as you've enabled through the virtual `GlobalInformation(QMN, Finder)` overload. YAGNI dictates that this additional overload should be removed..

Reply:
Sorry, I do not really follow what you think should be removed and also not why.

Reply:
I am really not sure what which method exactly we will not need.
However, let me elaborate on the overload of `GlobalInformation`. The idea behind it is not to provide different information by module, but to allow for performance enhancements in case it is cheaper to get the information relating to one module relative to getting everything. It is used in the overload of `DoGetInspectionResults` taking a QMN. That overload will become part of the `IInspection` in a future PR to allow running inspections for individual modules only.

Reply:
then my original point is moot :)